pkg/cmd/kind/test/internal/testcmd/testcmd.go

```diff
@@ -0,0 +1,372 @@
+	if f.IPFamily != string(v1alpha4.IPv4Family) && f.IPFamily != string(v1alpha4.IPv6Family) {
+		return fmt.Errorf(
+			"invalid --ipfamily %q; valid are: %q, %q",
+			f.IPFamily, v1alpha4.IPv4Family, v1alpha4.IPv6Family,
+		)
+	}
+	return nil
+}
+
+// Details represents injected details of a particular test command
+type Details struct {
+	Name       string
+	UsageLong  string
+	UsageShort string
+	Focus      string
+	Skip       string
+	Parallel   bool
+}
+
+// New returns a new cobra.Command for Kubernetes testing with ginkgo-e2e.sh
+func New(logger log.Logger, streams cmd.IOStreams, details Details) *cobra.Command {
+	flags := &flagpole{}
+	cmd := &cobra.Command{
+		Args:  cobra.NoArgs,
+		Use:   details.Name,
+		Short: details.UsageShort,
+		Long:  details.UsageLong,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// flag defaulting & validation
+			if err := flags.Process(); err != nil {
+				return err
+			}
+
+			// handle test command overrides
+			if focus, set := os.LookupEnv("KIND_TEST_FOCUS_OVERRIDE"); set {
+				details.Focus = focus
+				logger.Warnf("KIND_TEST_FOCUS_OVERRIDE overrode --ginkgo.focus to be: %s", focus)
+			}
+			if skip, set := os.LookupEnv("KIND_TEST_SKIP_OVERRIDE"); set {
+				details.Skip = skip
+				logger.Warnf("KIND_TEST_SKIP_OVERRIDE overrode --ginkgo.skip to be: %s", skip)
+			}
+
+			// actually run the command
+			return runGinkgo(logger, flags, details)
+		},
+	}
+	cmd.Flags().StringVar(&flags.IPFamily, "ip-family", string(v1alpha4.IPv4Family), "test cluster IP family")
+	cmd.Flags().StringVar(&flags.Name, "name", details.Name, "test cluster context name")
+	cmd.Flags().StringVar(&flags.Artifacts, "artifacts-dir", "", "the directory in which to output test results")
+	cmd.Flags().StringVar(&flags.ImageName, "image", "kindest/node:kubernetes-presubmit", "name:tag of the built kind node image")
+	return cmd
+}
+
+// TODO: junit output for steps a la kubetest
+func runGinkgo(logger log.Logger, flags *flagpole, details Details) (err error) {
+	// first debug the version we're using
+	logger.V(0).Infof("Testing with %s", version.DisplayVersion())
+
+	// then build
+	if err := buildWithGinkgoAndTests(logger, flags); err != nil {
+		return err
+	}
+
+	// setup shared provider for doing cluster things
+	provider := cluster.NewProvider(
+		cluster.ProviderWithLogger(logger),
+	)
+
+	// Check if the cluster already exists
+	if err := checkAlreadyExists(provider, flags.Name); err != nil {
+		return err
+	}
+
+	// then before we do anything more, arrange for cleanup to happen
+	cleanup := func(calledFromSignalHandler bool) {
+		err2 := provider.CollectLogs(flags.Name, filepath.Join(flags.Artifacts, "logs"))
+		err3 := provider.Delete(flags.Name, flags.kubeconfigPath())
+		// don't touch err if we're being called from another goroutine
+		// on program early exit, or if err is already non-nil
+		if calledFromSignalHandler || err != nil {
+			return
+		} else if err2 != nil {
+			err = err2
+		} else {
```

Comment:
this is only `if err3 != nil` or we alway assign err3 to err at this point?

Reply:
we don't care at this point, because if err3 is nil at this point so are err2 and err /shrug